### PR TITLE
music control

### PR DESCRIPTION
**Pull request by red031000**
moves the music control from gecko to pbrEngine itself, for finer control over what music we may/may not want enabled

### CODE REVIEW

File: pbrEngine/engine.py [hunk 1]
```diff
@@ -343,6 +345,9 @@ def _initDolphinWatch(self, watcher):
         self._newRng()  # avoid patterns. Unknown which patterns this avoids, if any.
         self._setState(EngineStates.INIT)
         self._lastInput = WiimoteButton.TWO  # to be able to click through the menu
+        if not self.musicEnabled:
+            self.disableMusic()
+            self._musicCurrentlyEnabled = False
```
Comment (Felk) [inline pbrEngine/engine.py:371]: else true?

File: pbrEngine/engine.py [hunk 2]
```diff
@@ -407,6 +412,9 @@ def reset(self):
 
         self._win_result_addr = None
         self._win_result = "draw"
+        if not self.musicEnabled and self._connected:
+            self.disableMusic()
+            self._musicCurrentlyEnabled = False
```
Comment (Felk) [inline pbrEngine/engine.py:441]: else true?

File: pbrEngine/engine.py [hunk 3]
```diff
@@ -691,6 +749,291 @@ def _resetBlur(self):
...
+        """
+        Enables the music for the Pokemon Selection screen. (System05)
+        """
+        self._enableSong(Locations.SONG_MON_SELECTION, "sound/System05.brstm")
+
+    def _enableControlsMusic(self):
+        """
+        Enables the music for the Controls screen. (System09)
+        """
+        self._enableSong(Locations.SONG_CONTROLS, "sound/System09.brstm")
```
Comment (Felk) [inline pbrEngine/engine.py]: do these need to have a separate method each? I suggest just putting them in a list, maybe tuples with their respective defaults

File: pbrEngine/engine.py [hunk 4]
```diff
@@ -691,6 +749,291 @@ def _resetBlur(self):
         self._dolphin.write32(Locations.BLUR1.value.addr, DefaultValues["BLUR1"])
         self._dolphin.write32(Locations.BLUR2.value.addr, DefaultValues["BLUR2"])
 
+    def _disableTitleMusic(self):
+        """
+        Disables the music on the title screen.
+        """
+        self._disableSong(Locations.SONG_TITLE, 6)
```
Comment (Felk) [inline pbrEngine/engine.py]: what does _6_ mean here? What do the others magic numbers mean?

File: pbrEngine/engine.py [hunk 5]
```diff
@@ -691,6 +749,291 @@ def _resetBlur(self):
...
+        Enables the music for the Controls screen. (System09)
+        """
+        self._enableSong(Locations.SONG_CONTROLS, "sound/System09.brstm")
+
+    def _enableBossMusic(self):
+        path = ""
+        side = None
+        if self.avatars["blue"]["APPEARANCE"]["CHARACTER_STYLE"] > 6:
+            side = "blue"
+        elif self.avatars["red"]["APPEARANCE"]["CHARACTER_STYLE"] > 6:
```
Comment (Felk) [inline pbrEngine/engine.py:748]: magic numbers

File: pbrEngine/engine.py [hunk 6]
```diff
@@ -691,6 +749,291 @@ def _resetBlur(self):
...
+            return
+        if self.avatars[side]["APPEARANCE"]["CHARACTER_STYLE"] in [7, 8]:
+            path = "sound/Battleboss01.brstm"
+        elif self.avatars[side]["APPEARANCE"]["CHARACTER_STYLE"] == 9:
+            path = "sound/Battleboss03.brstm"
+        elif self.avatars[side]["APPEARANCE"]["CHARACTER_STYLE"] == 10:
+            path = "sound/Battleboss04.brstm"
+        elif self.avatars[side]["APPEARANCE"]["CHARACTER_STYLE"] == 11:
+            path = "sound/Battleboss02.brstm"
+        elif self.avatars[side]["APPEARANCE"]["CHARACTER_STYLE"] == 12:
```
Comment (Felk) [inline pbrEngine/engine.py]: more magic numbers everywhere

Comment (Felk) [reply pbrEngine/engine.py]: you can extract stuff like `self.avatars[side]["APPEARANCE"]["CHARACTER_STYLE"]` to not have so many copy-paste accessors

File: pbrEngine/engine.py [hunk 7]
```diff
@@ -691,6 +749,291 @@ def _resetBlur(self):
...
+            path = "sound/Battlelastboss.brstm"
+
+        addrs = [Locations.SONG_WATERFALL, Locations.SONG_MAGMA, Locations.SONG_COURTYARD, Locations.SONG_SUNNY_PARK,
+                 Locations.SONG_SUNSET, Locations.SONG_STARGAZER, Locations.SONG_GATEWAY, Locations.SONG_NEON,
+                 Locations.SONG_MAIN_STREET, Locations.SONG_CRYSTAL, Locations.SONG_LAGOON]
+        for addr in addrs:
+            self._disableSong(addr, 8)
```
Comment (Felk) [inline pbrEngine/engine.py]: 8?

File: pbrEngine/engine.py [hunk 8]
```diff
@@ -691,6 +749,291 @@ def _resetBlur(self):
...
+            self._enableSong(addr, path)
```
Comment (Felk) [inline pbrEngine/engine.py:769]: what does this loop do? just disable and enable everything?

File: pbrEngine/engine.py [hunk 9]
```diff
@@ -691,6 +749,291 @@ def _resetBlur(self):
...
+
+        self._disableSong(Locations.SONG_FANFARE_VAR1, 7)  # set boss fanfares
+        self._enableSong(Locations.SONG_FANFARE_VAR1, "sound/ME_Fan02.brstm")
+        self._disableSong(Locations.SONG_FANFARE_VAR3, 7)
+        self._enableSong(Locations.SONG_FANFARE_VAR3, "sound/ME_Fan04.brstm")
+        self._disableSong(Locations.SONG_FANFARE_VAR5, 7)
+        self._enableSong(Locations.SONG_FANFARE_VAR5, "sonnd/ME_Fan06.brstm")
```
Comment (Felk) [inline pbrEngine/engine.py:776]: more magic numbers

File: pbrEngine/engine.py [hunk 10]
```diff
@@ -691,6 +749,291 @@ def _resetBlur(self):
...
+
+    def _disableSong(self, loc, count):
+        writes = []
+        for i in range(count):
+            writes.append((32, loc.value.addr + 0x4 * i, 0))
+        self._dolphinIO.writeMulti(writes)
```
Comment (Felk) [inline pbrEngine/engine.py:782]: oh, those magic numbers are "count"? please pass the arguments as keywords for clarity

Comment (Felk) [reply pbrEngine/engine.py:782]: why 0x4? what does the loop do? short explanation

File: pbrEngine/engine.py [hunk 11]
```diff
@@ -691,6 +749,291 @@ def _resetBlur(self):
...
+
+    def _enableSong(self, loc, path):
+        bytes = path.encode()
+        for i, byte in enumerate(bytes):
+            self._dolphin.write(8, loc.value.addr + i, byte)
```
Comment (Felk) [inline pbrEngine/engine.py]: also a short explanation please

File: pbrEngine/engine.py [hunk 12]
```diff
@@ -1177,6 +1520,11 @@ def _matchStart(self):
         '''
         logger.info("Starting PBR match")
         self._injectAvatars()
+        if self._musicCurrentlyEnabled:
+            self.enableMusic()
+            if (self.avatars["blue"]["APPEARANCE"]["CHARACTER_STYLE"] > 6 or
+                    self.avatars["red"]["APPEARANCE"]["CHARACTER_STYLE"] > 6):
+                self._enableBossMusic()
```
Comment (Felk) [inline pbrEngine/engine.py]: more magic numbers

File: pbrEngine/engine.py [hunk 13]
```diff
@@ -198,6 +198,8 @@ def _crash(reason=None):
 
         self._matchVolume = 100
         self._fMatchAnnouncer = True
+        self.musicEnabled = False
+        self._musicCurrentlyEnabled = False
```
Comment (Felk) [inline pbrEngine/engine.py]: how are these two variables used? what two of them?

**Review by Felk — CHANGES_REQUESTED** [review]

File: pbrEngine/engine.py (hunk 13 above)
Comment (red031000) [reply pbrEngine/engine.py]: musicEnabled is set by the core to signify whether music should be enabled every match, _musicCurrentlyEnabled is whether the current match has music

**Review by Felk — APPROVED** [review]

**Pull request merged**